### ISSUE TITLE
.rate function issue with  negative/positive values switches

### ISSUE THREAD

**Issue by freekatet** [processing/p5.js-sound#236]
Hello,

playing files using a negative value for the rate doesn't react as it should .

By changing the range of the rate between -2 to 2 for example, i noticed that as soon i change the rate value to a negative value it plays in reverse but starting at the end of the audio file, if you change back the rate to a positive value it jump back where the file was playing before giving a negative value.

Example: 
play at rate 1 for 30 sec, 
changing the rate to -1 now plays from 0 sec backward for 15 sec
 putting back rate at 1 jump back at 30sec in the audio file, 
putting -1 again jump back at -15sec
etc..

it happen in any browsers i tried, computer/mobile.

its like if the code was playing the file 2 times one for positive rate , the other for negative rate and they both keep their own timing (making impossible to scratch like on a turntable for example).

thanks!

**Comment by therewasaguy** [comment]
Thanks for the issue, @freekatet!

Started work on this here https://github.com/therewasaguy/p5.js-sound/tree/negative-playbackRate

You can use this file if you'd like to try it out https://raw.githubusercontent.com/therewasaguy/p5.js-sound/negative-playbackRate/lib/p5.sound.js

**Comment by freekatet** [comment]
@therewasaguy thanks that's great! it's working as expected now on my side. there is still a little strange behavior when you switch positive/negative too fast ( i got a blank for a short time) 
i will do some test with lerp to see if it helps..

thanks a lot!

**Comment by therewasaguy** [comment]
thanks @freekatet, can you share a code snippet?

**Comment by freekatet** [comment]
@therewasaguy sure! i actually just did the test on the playbackRate example and changed the rate on the sketch

you can try here 
http://halftheory.com/freeka_tet/5/1/playbackRate/

if you move you mouse slowly its fine but as soon you go faster you will see what i mean .

ps: going too fast crash the player

**Comment by therewasaguy** [comment]
hmm, I am seeing that if you move really fast, it silences the soundfile output. Is that what you mean by crashing the player? That's what I was just seeing with your example, and when the console and set `soundFile.amp(1),` I could hear it again. Not totally sure how that's happening but it must have to do with the way we are trying to do a quick fade out then fade back in as part of `reverseBuffer` to avoid a digital click. I'll look at this more soon!

**Comment by freekatet** [comment]
@therewasaguy yes that might be one of the problem, but not only, if you focus on the rate value being displayed, you will see that the value is freezing for a little time when you switch from negative to positive or opposite. ( example: value will freeze at 0.54 and then jump to -0.7 ) 
the sound will act exactly as the value displayed (same freeze)..

**Comment by therewasaguy** [comment]
ahh, you're working with a longer sound file than the original example, and the entire buffer is stored in memory. When we switch between positive and negative playback rates, we reverse the entire buffer, which takes a long time

```
       for (var i = 0; i < this.buffer.numberOfChannels; i++) {
          Array.prototype.reverse.call( this.buffer.getChannelData(i) );
        }
```
via https://github.com/processing/p5.js-sound/pull/241/files#diff-635fd3ddb6f066be99025f1994db3e30R1037

The buffer has 44100 samples per second, per channel, so even just a one minute buffer is pretty large.

Maybe there is a faster way to do this, but I'm not very optimistic that we'll be able to get around this issue.

**Comment by GoToLoop** [comment]
Given method **reverse()** already acts _in situ_:
https://Developer.Mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/reverse

There's no need for this **Array.prototype.reverse.call()** complex approach, right? How about this 1?
```js
      var buf = this.buffer, idx = buf.numberOfChannels;
      while (idx--)  buf.getChannelData(idx).reverse();

**Comment by GoToLoop** [comment]
Apparently, for most cases, it seems like _numberOfChannels_ is either 1 or 2 (either mono or stereo).
If that's a correct supposition, we don't need a loop to deal w/ just 2 objects:
```js
      var buf = this.buffer, channels = buf.numberOfChannels;
      buf.getChannelData(0).reverse();
      channels > 1 && buf.getChannelData(1).reverse();
```

**Comment by GoToLoop** [comment]
How about having an extra __*AudioBuffer*__ w/ its channels already reversed?

And within **reverseBuffer()**, rather than manually **reverse()** everything over & over, just assign to **_AudioBufferSourceNode_**::_buffer_ the current switched __*AudioBuffer*__:

https://Developer.Mozilla.org/en-US/docs/Web/API/AudioBufferSourceNode/buffer

**Comment by therewasaguy** [comment]
Great thinking, @GoToLoop! Yes, the typed array `reverse` method seems to complete in 1/10th the time. It must not've existed in 2014 when this function was made, but it looks safe to use now going back many browser versions. This change [here](https://github.com/processing/p5.js-sound/pull/241/commits/30eed31b81f1ca849995eaad0310cb86cb0fe970#diff-635fd3ddb6f066be99025f1994db3e30R1029) makes a big difference.

I also found that (re-)populating the `counterBuffer` (which keeps track of the current position in the track, and needs to be re-generated when the buffer changes until we can find a better way to handle this) was slow and was able to optimize that [here](https://github.com/processing/p5.js-sound/pull/241/commits/30eed31b81f1ca849995eaad0310cb86cb0fe970#diff-635fd3ddb6f066be99025f1994db3e30R1193). The counterBuffer is the same length as one channel of the regular audio buffer, and it's used to keep track of the current position in the playback buffer using the scriptProcessorNode, which has been deprecated, so this is a potential performance issue and something that'll have to be migrated to ScriptProcessor's replacement, the AudioWorker when it finally arrives (soon!)

Storing an extra AudioBuffer is an interesting idea! I tried it, but it didn't seem to impact performance, and it's probably best to save the memory since we're already taking up more memory than we need to just to play each audio sample _and_ keep track of the current time.

@freekatet let's see how the lib sounds for you with this https://github.com/therewasaguy/p5.js-sound/blob/94b0b1fd72699005e3b044e5655303596fe7618d/lib/p5.sound.js I was testing on a 4 minute audio file and it sounded smooth!

**Comment by freekatet** [comment]
@therewasaguy @GoToLoop thanks Guys! works awesome on my side on my laptop. 
For some reason its still acting weird on mobile browsing tho. any ideas? 
(i change the .min.js too and erase my web data)

**Comment by therewasaguy** [comment]
ah, that's probably the audio buffer size putting a strain on mobile browsers memory...is it any better with a smaller audio file? Not sure we can do any better than this, but if anyone wants to give it a shot that'd be fantastic!

**Comment by freekatet** [comment]
@therewasaguy sorry for not being more clear :/ 
when i mentioned the mobile version, i meant its the exact same problem as when i started this thread. i doubt it would be because of the audio buffer then ? 
thanks!

**Comment by therewasaguy** [comment]
hmm, will have to look at this again soon. Thanks for your patience and help debugging!

**Comment by freekatet** [comment]
Oh sure no problem, thanks for helping !

**Comment by MichaelPaulukonis** [comment]
Not sure if this is the same issue or not, but when I attempt to switch from a positive to a negative `rate`, playback apparently stops, UNLESS I set the negative rate TWICE.

-----

**Expected behavior**: playback with a positive rate when set to a negative rate should continue to play, at the same position, only in reverse.

**Actual behavior**: playback with a positive rate when set to a negative rate STOPS playback (unless the negative rate is set TWICE in succession)

-----



I have a short example @ https://editor.p5js.org/MichaelPaulukonis/sketches/T2EVbCom8

- type `r` to reverse direction (playback speed is logged to the console)
- select `once` or `twice` to choose how often the rate is set when reversing


The stopping effect does not occur when switching from negative to positive.
Setting the negative rate more than twice seems to have no additional effects.

I have not investigated the code as to why this may happen, and have no patch available at this time.
